File: cpp/graph/dinic.cpp
#include <bits/stdc++.h>
using namespace std;
template <class C>
struct Dinic
{
  struct E
  {
    int to, rev;
    C cap;
  };
  int n;
  vector<vector<E>> g;
  vector<int> lvl, it;
  Dinic(int n = 0) : n(n), g(n) {}
  void add_edge(int v, int to, C cap)
  {
    E a{to, (int)g[to].size(), cap}, b{v, (int)g[v].size(), 0};
    g[v].push_back(a);
    g[to].push_back(b);
  }
  bool bfs(int s, int t)
  {
    lvl.assign(n, -1);
    queue<int> q;
    lvl[s] = 0;
    q.push(s);
    while (!q.empty())
    {
      int v = q.front();
      q.pop();
      for (auto &e : g[v])
        if (e.cap > 0 && lvl[e.to] < 0)
          lvl[e.to] = lvl[v] + 1, q.push(e.to);
    }
    return lvl[t] >= 0;
  }
  C dfs(int v, int t, C f)
  {
    if (v == t)
      return f;
    for (int &i = it[v]; i < (int)g[v].size(); i++)
    {
      E &e = g[v][i];
      if (e.cap > 0 && lvl[e.to] == lvl[v] + 1)
      {
        C ret = dfs(e.to, t, min(f, e.cap));
        if (ret)
        {
          e.cap -= ret;
          g[e.to][e.rev].cap += ret;
          return ret;
        }
      }
    }
    return 0;
  }
  C flow(int s, int t)
  {
    C ans = 0, INF = numeric_limits<C>::max() / 4;
    while (bfs(s, t))
    {
      it.assign(n, 0);
      while (C f = dfs(s, t, INF))
        ans += f;
    }
    return ans;
  }
};

File: python/graph/dinic.py
from collections import deque

INF = 10**30


class Dinic:
    def __init__(self, n):
        self.n = n
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v, cap):
        self.g[u].append([v, cap, len(self.g[v])])
        self.g[v].append([u, 0, len(self.g[u]) - 1])

    def max_flow(self, s, t):
        flow = 0
        while True:
            level = [-1] * self.n
            q = deque([s])
            level[s] = 0
            while q:
                u = q.popleft()
                for v, cap, _ in self.g[u]:
                    if cap and level[v] < 0:
                        level[v] = level[u] + 1
                        q.append(v)
            if level[t] < 0:
                return flow
            it = [0] * self.n

            def dfs(u, f):
                if u == t:
                    return f
                for i in range(it[u], len(self.g[u])):
                    it[u] = i
                    v, cap, rev = self.g[u][i]
                    if cap and level[v] == level[u] + 1:
                        ret = dfs(v, min(f, cap))
                        if ret:
                            self.g[u][i][1] -= ret
                            self.g[v][rev][1] += ret
                            return ret
                return 0

            while True:
                pushed = dfs(s, INF)
                if not pushed:
                    break
                flow += pushed
